### ISSUE TITLE
Hardware assisted UART -> USB forwarding

### ISSUE THREAD

**Issue by jesultra** [lowRISC/opentitan#16549]
Before OpenTitan, Google has been using internally developed Security Chips, and has recently chosen to use Rust and TockOS as basis for the firmware on those chips.  TockOS discourages performing device driver logic in interrupt handler context, as doing so necessarily means that Rust security guarantees cannot be verified by the compiler.  Most device drivers use "bottom half" handlers exclusively, which run in ordinary kernel context, but can suffer from significant latency compared to actual interrupt handlers.  On the previous Google Security Chip, only two device drivers needed to rely on low-latency interrupt handling for their correctness: the I2C target driver, and the UART driver.

OpenTitan I2C hardware has clock stretching features, and ability to recover transaction boundaries through its acq_fifo, meaning that on OpenTitan, the Google I2C driver does not need low-latency handling.  This leaves only the UART driver.  We further worry that flash erase/write for instance could cause latency even for interrupt context code, and this amplifies the desire to look for ways of eliminating the need for low-latency processing of the OpenTitan UARTs.

The single task that requires low-latency processing is forwarding of incoming UART data to a particular USB device endpoint,  such that an external debugger (SuzyQ) can be used to extract logging info from other chips in the device (AP, EC, FPMCU).

I realize that what I am asking here is a major and invasive feature, and I should probably have asked a year ago, but I want to put it in writing now anyway, since the potential benefit is significant.

### Feature request
The UART IP will have a new "USB RX mode" bit, which controls whether incoming complete bytes from the hardware shift register go into the RX FIFO as they currently do, or instead go directly into one of the 32 buffers of the USB IP.
When enabled, two of the USB buffers are reserved for use by this UART, and should not be accessed by the CPU.  (The allocation could be static, that is e.g. UART0 can only use buffer 0 and buffer 1, UART1 can only use buffer 2 and buffer 3, etc.)
When enabled, the UART IP will directly manipulate the INCFG[..] control signals to a particular endpoint (allowing a configurable endpoint would be nice, but static allocation would be OK as well,) in such a way that once either a 64-byte USB buffer has become full, or the UART RX timeout expires while there are some characters in one 64-byte buffer and the other 64-byte buffer is empty, then the RDY signal is asserted to the USBDEV IP.

There are many details to the proposal, which need to be worked out.

**Comment by mdhayter** [comment]
It may be possible to use the UART fifos instead of the USB buffer. More like the way the "usbuart" is done. The fifos would be presented as an additional endpoint. There would need to be a bit of work done to make the number of edpoints seen by the bottom half of the usb be more than the stuff that controls the buffer, but I think that would be less tricky than having to find a way to get the UART read/write to the buffer.

At least that was my thinking two years ago when I briefly considered this.

**Comment by a-will** [comment]
Correct me if I'm wrong, but this isn't a true low latency requirement, is it? It seems like a simpler approach might be to increase the size of the UART buffers to cover the deferred work latency, for example.

If we were also concerned about CPU utilization for moving data, then we might need to think about accelerators, but it wasn't mentioned in the rationale.

**Comment by cdgori** [comment]
This sort of "proto-DMA" would need to be really carefully considered because I think it might break some of the assumptions we made long ago about how the fabric would be configured and which endpoints were able to be hosts.  If the increased buffer solution was adequate to cover the deferred work that seems preferable.

I'm not saying the hardware assisted forwarding wouldn't work, we would just have to re-check our assumptions re: IOPMPs and some other constructions.  We are likely going to have to re-check them anyway for the integrated use case.

Tagging @tjaychen since he was part of that earlier discussion about host/device splits and the various steps we took to protect the fabric, I can't remember who else was involved in discussion that but he might.

**Comment by tjaychen** [comment]
hey all,
so a lot to say here. 

I think as @a-will points out, the thing we probably want to do in the short term is to increase the UART buffer size. We probably need to do some quick math as to what would happen if a UART deposit happens to collide with a flash operation at typical speeds (i assume 115K)?

The longer term solution, as @cdgori points out, really is a DMA.  We allow the DMA to move the data directly for us.  We have long planned for something like this, which is why we requested the lowrisc folks (@GregAC @vogelpi ) to make their IOPMP something we can instantiate in the open.  We would do it for anything that has host capability.  But more importantly, we have other partners right now who are looking into DMAs for a different use case.  

So in summary, if this is a concern, let's increase the buffer size as Alex suggests for now.  Then we can give it a couple of months for the DMA story to fully shake out and just plug that in as a more generic solution to issues like this in general.

**Comment by GregAC** [comment]
I'd certainly prefer to just increase the UART buffer size over other solutions for earlgrey. If we were to implement this I'd favour a direct, special purpose, UART -> USB channel rather than allowing the UART device to issues TL transactions onto the interconnect and even that would be a reasonable chunk of work. At this stage we need a strong reason to consider implementing it.

**Comment by a-will** [comment]
At 115.2k baud (8n1), half the size of the current UART FIFO (16 B) would take ~1.39 ms at full utilization. Is the latency expected to the first read expected to be that long?

**Comment by cdgori** [comment]
> At 115.2k baud (8n1), half the size of the current UART FIFO (16 B) would take ~1.39 ms at full utilization. Is the latency to the first read expected to be that long?

If I understand the concern properly, at least one issue raised by @jesultra is the max latency of the flash erase/write process, perhaps @tjaychen has an idea of the latency there and therefore how much we might need to increase the UART FIFO size to handle that case.

I would strongly prefer to avoid changing the TL initiator pathways at this stage, but if we ultimately decide to do so, I would also strongly favor a solution like @GregAC proposes, where the UART has a separate/isolated path to deposit data to the USB - allowing the UART to directly drive TL transactions to the interconnect would require even more analysis/thought about the implications.  We can, however, take that under advisement for the integrated use case and do that analysis there.

**Comment by tjaychen** [comment]
@jesultra can you have a look at the latency point @a-will raised? I think if we would like to do anything, increasing the buffer size is the likely choice, but we would need to know soon.

**Comment by moidx** [comment]
Marking as P1 to track the resolution of this discussion.

**Comment by GregAC** [comment]
We need to get a resolution to this for M2.5 with a strong preference towards just increasing UART buffer size vs introducing the suggested hardware forwarding feature.

**Comment by GregAC** [comment]
@jesultra will the large buffer sizes proposed be sufficient to address your concerns? At this point I'm not sure the proposed forwarding path is feasible for M2.5

**Comment by jesultra** [comment]
On the previous security chip, a 64 byte UART RX FIFO is enough that our interrupt handler does not lose characters. This has the disadvantage of circumventing some Rust compiler checking, but we can live with that on OpenTitan as well 
That the same FIFO size will work is conditional on some things, though. The previous chip allowed code execution during flash erase/write (as long as the two took place in opposite halves of the memory), does OpenTitan do the same? If not, it could take several thousand bytes of RX FIFO to queue all the data arriving during flash writes.

**Comment by a-will** [comment]
> The previous chip allowed code execution during flash erase/write (as long as the two took place in opposite halves of the memory), does OpenTitan do the same? If not, it could take several thousand bytes of RX FIFO to queue all the data arriving during flash writes.

Yes, the Earl Grey design has two independent flash banks, with the instruction fetch as a separate read requester from the flash controller.

This is my understanding from the current docs and RTL. I wasn't readily able to determine if we check this specific timing in DV, though.

**Comment by marnovandermaas** [comment]
@alees24 can you estimate this as part of the USB dev M2.5 effort estimation?

**Comment by GregAC** [comment]
As this point I think the only feasible option is increasing the UART FIFO depth (or some other solution if there are any). In particular adding a direct forwarding path to USB from the UART isn't possible for M2.5 so it doesn't need inclusion in the USB estimate.

**Comment by marnovandermaas** [comment]
Captured in UART estimate for M2.5:
> estimate 8

**Comment by cdgori** [comment]
> As this point I think the only feasible option is increasing the UART FIFO depth (or some other solution if there are any). In particular adding a direct forwarding path to USB from the UART isn't possible for M2.5 so it doesn't need inclusion in the USB estimate.

Just to be 100% clear: if for any reason the idea of the direct forwarding path from UART to USB is revived, it will need (potentially substantial) security review, as it potentially changes some assumptions we made.

Increasing FIFO depth does not have that requirement :)

**Comment by moidx** [comment]
Leaving as P1 as to address UART FIFO updates.

**Comment by GregAC** [comment]
I've started worked on the expansion, PR here: https://github.com/lowRISC/opentitan/pull/17829

We need to agree:

- If 64 is the appropriate size to expand to
- What the new watermark levels should be

**Comment by alees24** [comment]
A few points

- The USB device supports a max packet size of 64 bytes and this gives peak throughput on the USB.
- It would be best if the software could avoid transferring data twice, ie. if it can pick up 64 bytes as a chunk and write them into the USB packet buffer directly (essentially specialized_memcpy(usb device addr, uart device, 64)) if we're really concerned about throughput and the impact of software latency
- The UART will continue receiving into its FIFO during the time between 64 bytes being available in the FIFO and the point at which software has responded to the interrupt and started shuffling the 64-byte payload to the USB device

Perhaps we should aim to set the UART FIFO size to 64 bytes plus (anticipated worst case latency / data rate into the UART) and then the software should be architected to try always to shuffle 64 byte lumps? (Obviously it also must not wait too long in its efforts to gather 64 bytes)

**Comment by jesultra** [comment]
I think I understand the proposal, and that is how one would ordinarily get fast performance, by having the UART interrupt handler copy data from the UART FIFO (which should be 128 bytes at least), into the USB endpoint ram.  (Triggered by 64-byte threshold interrupt, or by something like "ten character times elapsed without any UART input".)

My problem arise from the fact that we are using TockOS, which puts some limitations on what can be done from interrupt handlers.  I do not think we will be able to access our USB driver from UART interrupt handler, at least, that was our tradeoff on our current security chip, and we plan using the same code base.  (Accessing USB driver from true interrupt would mean lots of more Rust-unsafe operations in the USB driver.)

Basically, in order to reap benefit from Rust's compiler, we need to have as little as possible take place in the true interrupt handlers. Currently the UART RX interrupt simply copies data from the FIFO into a cyclic buffer in non-USB RAM, and then a Rust bottom half handler copies from that RAM buffer to the usb endpoint buffer.  I think we want to do the same on OpenTitan.

**Comment by johngt** [comment]
Updated FIFO size & watermark captured in linked PR from @GregAC - https://github.com/lowRISC/opentitan/pull/17829
